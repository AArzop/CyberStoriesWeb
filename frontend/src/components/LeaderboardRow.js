import React, { Component } from 'react'
import withStyles from 'react-jss'
import classNames from 'classnames'
import { compose } from 'redux'
import { NavLink } from 'react-router-dom'
import { lerpColor3 } from '../utils/color'

const gbstyles = {
  growingBar: {
    height: '14px',
    borderRadius: '7px'
  },
  wrapper: {
    display: 'flex',
    flexDirection: 'row'
  }
}

class GrowingBar extends Component {
  constructor (props) {
    super(props)
    this.state = {
      width: 0
    }

    this.maxScore = 100

    this.grow = this.grow.bind(this)
  }

  componentWillReceiveProps (nextProps, nextContext) {
    this.grow(nextProps.value)
  }

  componentDidMount () {
    this.grow(this.props.value)
  }

  grow (value) {
    let count = 0
    const intervalId = setInterval(() => {
      if (this.state.width < value) {
        this.setState(state => ({
          width: state.width + 1
        }))
      }
      if (count++ >= value) {
        window.clearInterval(intervalId)
      }
    }, 1)
  }

  render () {
    const color = lerpColor3('#ff2a2a', '#ffaa00', '#69d453', this.state.width / 100)
    return (
      <div className={classNames(this.props.classes.wrapper, this.props.className)}>
        <div
          className={classNames(
            'm-1',
            this.props.classes.growingBar
          )}
          style={{
            width: this.state.width,
            backgroundColor: color
          }}
        />
        <div
          style={{
            color: color,
            fontWeight: 'bold'
          }}
        >{this.state.width} %
        </div>
      </div>
    )
  }
}

GrowingBar = withStyles(gbstyles)(GrowingBar)

const styles = {
  bar: {
    marginLeft: '100px'
  },
  pseudo: {
    cursor: 'pointer',
    color: 'grey',
    '&:hover': {
      textDecoration: 'none',
      fontWeight: 'bold',
      color: 'grey'
    }
  },
  cell: {
    borderBottom: '1px solid rgba(100, 100, 100, 0.3)'
  }
}

class LeaderboardRow extends Component {
  render () {
    const entry = this.props.entry

    const globalScore = this.props.globalScore

    return (
      <div className={'row'}>
        <div
          className={classNames('col-sm p-1', this.props.classes.cell)}>
          <NavLink to={`/leaderboard/${entry.id}/`} className={this.props.classes.pseudo}>
            {entry.pseudo}
          </NavLink>
        </div>
        <div className={classNames('col-sm p-1', this.props.classes.cell)}>
          <GrowingBar value={globalScore(entry)} className={this.props.classes.bar}/>
        </div>
        <div className={classNames('col-sm p-1', this.props.classes.cell)} data-test-class={'numberOfGamesCell'}>
          {entry.gameResults.length > 10 ? '10+' : entry.gameResults.length}
        </div>
      </div>
    )
  }
}

LeaderboardRow.propTypes = {}

LeaderboardRow.defaultProps = {
  entry: {},
  globalScore: () => 0
}

export default compose(
  withStyles(styles)
)(LeaderboardRow)
