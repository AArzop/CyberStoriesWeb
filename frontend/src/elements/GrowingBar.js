import React, {Component} from "react";
import classNames from "classnames";
import withStyles from "react-jss";
import {lerpColor, lerpColor3} from "../utils/color";

const styles = {
    growingBar: {
        height: '14px',
        borderRadius: '7px'
    },
    wrapper: {
        display: 'flex',
        flexDirection: 'row'
    }
};

class GrowingBar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            width: 0
        };

        this.maxScore = 100;

        this.grow = this.grow.bind(this);
    }

    componentWillReceiveProps(nextProps, nextContext) {
        this.grow(nextProps.value);
    }

    componentDidMount() {
        this.grow(this.props.value);
    }

    grow(value) {
        let count = 0;
        const intervalId = setInterval(() => {
            if (this.state.width < value) {
                this.setState(state => ({
                    width: state.width + 1
                }));
            }
            if (count++ >= value) {
                window.clearInterval(intervalId)
            }
        }, 1);
    }

    render() {
        const color = lerpColor3('#ff2a2a', '#ffaa00', '#69d453', this.state.width / 100);
        return (
            <div className={classNames(this.props.classes.wrapper, this.props.className)}>
                <div
                    className={classNames(
                        "m-1",
                        this.props.classes.growingBar
                    )}
                    style={{
                        width: this.state.width,
                        backgroundColor: color
                    }}
                />
                <div
                    style={{
                        color: color,
                        fontWeight: 'bold'
                    }}
                >{this.state.width} %
                </div>
            </div>
        );
    }
}

export default withStyles(styles)(GrowingBar);
