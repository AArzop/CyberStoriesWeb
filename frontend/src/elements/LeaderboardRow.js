import React, { Component } from 'react'
import withStyles from 'react-jss'
import classNames from 'classnames'
import { compose } from 'redux'
import { NavLink } from 'react-router-dom'
import GrowingBar from './GrowingBar'

const styles = {
  bar: {
    marginLeft: '100px'
  },
  pseudo: {
    cursor: 'pointer',
    color: 'grey',
    '&:hover': {
      textDecoration: 'none',
      fontWeight: 'bold',
      color: 'grey'
    }
  },
  cell: {
    borderBottom: '1px solid rgba(100, 100, 100, 0.3)'
  }
}

class LeaderboardRow extends Component {
  render () {
    const entry = this.props.entry

    const globalScore = this.props.globalScore

    return (
      <div className={'row'}>
        <div
          className={classNames('col-sm p-1', this.props.classes.cell)}>
          <NavLink to={`/leaderboard/${entry.id}/`} className={this.props.classes.pseudo}>
            {entry.pseudo}
          </NavLink>
        </div>
        <div className={classNames('col-sm p-1', this.props.classes.cell)}>
          <GrowingBar value={globalScore(entry)} className={this.props.classes.bar}/>
        </div>
        <div className={classNames('col-sm p-1', this.props.classes.cell)} data-test-class={'numberOfGamesCell'}>
          {entry.gameResults.length > 10 ? '10+' : entry.gameResults.length}
        </div>
      </div>
    )
  }
}

LeaderboardRow.propTypes = {}

LeaderboardRow.defaultProps = {
  entry: {},
  globalScore: () => 0
}

export default compose(
  withStyles(styles)
)(LeaderboardRow)
