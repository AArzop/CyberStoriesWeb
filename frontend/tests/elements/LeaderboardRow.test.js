import LeaderboardRow from '../../src/elements/LeaderboardRow'
import React from 'react'
import { MemoryRouter } from 'react-router'
import { NavLink } from 'react-router-dom'
import renderer from 'react-test-renderer'
import GrowingBar from '../../src/elements/GrowingBar'

test('if leaderboard row prints right data from props', () => {
  const entryMock = {
    id: 0,
    pseudo: 'John',
    email: 'john.doe@gmail.com',
    gameResults: [
      {
        id: 0,
        date: '2019/05/01',
        stage1Score: 10,
        stage2Score: 20,
        stage3Score: 30,
        stage4Score: 40,
        visible: true,
        player: 0
      }
    ]
  }

  const globalScoreMock = () => 1234

  const component = renderer.create(
    <MemoryRouter>
      <LeaderboardRow
        entry={entryMock}
        globalScore={globalScoreMock}
      />
    </MemoryRouter>
  )

  const json = component.toJSON()
  expect(json).toMatchSnapshot()

  const instance = component.root

  const navlink = instance.findByType(NavLink)
  expect(navlink.props.children).toBe(entryMock.pseudo)

  const growingBar = instance.findByType(GrowingBar)
  expect(growingBar.props.value).toBe(globalScoreMock(entryMock))

  const numberOfGamesCell = instance.findByProps({ 'data-test-class': 'numberOfGamesCell' })
  expect(numberOfGamesCell.props.children).toBe(
    entryMock.gameResults.length > 10 ? '10+' : entryMock.gameResults.length
  )
})
