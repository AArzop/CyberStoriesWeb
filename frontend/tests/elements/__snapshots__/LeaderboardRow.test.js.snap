// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if leaderboard row prints right data from props 1`] = `
<div
  className="row"
>
  <div
    className="col-sm p-1 LeaderboardRow-cell-0-1-3"
  >
    <a
      aria-current={null}
      className="LeaderboardRow-pseudo-0-1-2"
      href="/leaderboard/0/"
      onClick={[Function]}
    >
      John
    </a>
  </div>
  <div
    className="col-sm p-1 LeaderboardRow-cell-0-1-3"
  >
    <div
      className="GrowingBar-wrapper-0-1-5 LeaderboardRow-bar-0-1-1"
    >
      <div
        className="m-1 GrowingBar-growingBar-0-1-4"
        style={
          Object {
            "backgroundColor": "#ff2a2a",
            "width": 0,
          }
        }
      />
      <div
        style={
          Object {
            "color": "#ff2a2a",
            "fontWeight": "bold",
          }
        }
      >
        0
         %
      </div>
    </div>
  </div>
  <div
    className="col-sm p-1 LeaderboardRow-cell-0-1-3"
    data-test-class="numberOfGamesCell"
  >
    1
  </div>
</div>
`;
